File: plugins/methods/src/com/intellij/methods/fs/MethodsVirtualFileSystem.java
package com.intellij.methods.fs;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileListener;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.intellij.openapi.vfs.VirtualFileSystem;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class MethodsVirtualFileSystem extends VirtualFileSystem {
  public static final String PROTOCOL = "methods";

  private final Map<String, MethodsVirtualFile> myFiles = new ConcurrentHashMap<>();

  public static MethodsVirtualFileSystem getInstance() {
    return (MethodsVirtualFileSystem)VirtualFileManager.getInstance().getFileSystem(PROTOCOL);
  }

  @Override
  public @NotNull String getProtocol() {
    return PROTOCOL;
  }

  public @NotNull MethodsVirtualFile createFile(@NotNull String path, @NotNull String content) {
    MethodsVirtualFile file = new MethodsVirtualFile(path, content);
    myFiles.put(path, file);
    return file;
  }

  @Override
  public @Nullable MethodsVirtualFile findFileByPath(@NotNull String path) {
    return myFiles.get(path);
  }

  @Override
  public void refresh(boolean asynchronous) {
  }

  @Override
  public @Nullable VirtualFile refreshAndFindFileByPath(@NotNull String path) {
    return findFileByPath(path);
  }

  @Override
  public void addVirtualFileListener(@NotNull VirtualFileListener listener) {
  }

  @Override
  public void removeVirtualFileListener(@NotNull VirtualFileListener listener) {
  }

  @Override
  protected void deleteFile(Object requestor, @NotNull VirtualFile vFile) {
    throw new IncorrectOperationException();
  }

  @Override
  protected void moveFile(Object requestor, @NotNull VirtualFile vFile, @NotNull VirtualFile newParent) {
    throw new IncorrectOperationException();
  }

  @Override
  protected void renameFile(Object requestor, @NotNull VirtualFile vFile, @NotNull String newName) {
    throw new IncorrectOperationException();
  }

  @Override
  protected @NotNull VirtualFile createChildFile(Object requestor, @NotNull VirtualFile vDir, @NotNull String fileName) {
    throw new IncorrectOperationException();
  }

  @Override
  protected @NotNull VirtualFile createChildDirectory(Object requestor, @NotNull VirtualFile vDir, @NotNull String dirName) {
    throw new IncorrectOperationException();
  }

  @Override
  protected @NotNull VirtualFile copyFile(Object requestor, @NotNull VirtualFile virtualFile, @NotNull VirtualFile newParent, @NotNull String copyName) {
    throw new IncorrectOperationException();
  }

  @Override
  public boolean isReadOnly() {
    return true;
  }
}

File: plugins/methods/src/com/intellij/methods/fs/MethodsVirtualFileSystem.kt
package com.intellij.methods.fs

import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileListener
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.openapi.vfs.VirtualFileSystem
import com.intellij.util.IncorrectOperationException
import org.jetbrains.annotations.NonNls

import java.util.concurrent.ConcurrentHashMap

/**
 * In-memory VFS for synthetic method listings.
 */
class MethodsVirtualFileSystem : VirtualFileSystem() {

    private val files = ConcurrentHashMap<String, MethodsVirtualFile>()

    override fun getProtocol(): String = PROTOCOL

    fun createFile(path: String, content: String): MethodsVirtualFile {
        val file = MethodsVirtualFile(path, content)
        files[path] = file
        return file
    }

    override fun findFileByPath(path: String): MethodsVirtualFile? = files[path]

    override fun refresh(asynchronous: Boolean) {}

    override fun refreshAndFindFileByPath(path: String): VirtualFile? = findFileByPath(path)

    override fun addVirtualFileListener(listener: VirtualFileListener) {}

    override fun removeVirtualFileListener(listener: VirtualFileListener) {}

    override fun deleteFile(requestor: Any?, vFile: VirtualFile) {
        throw IncorrectOperationException()
    }

    override fun moveFile(requestor: Any?, vFile: VirtualFile, newParent: VirtualFile) {
        throw IncorrectOperationException()
    }

    override fun renameFile(requestor: Any?, vFile: VirtualFile, newName: String) {
        throw IncorrectOperationException()
    }

    override fun createChildFile(requestor: Any?, vDir: VirtualFile, fileName: String): VirtualFile {
        throw IncorrectOperationException()
    }

    override fun createChildDirectory(requestor: Any?, vDir: VirtualFile, dirName: String): VirtualFile {
        throw IncorrectOperationException()
    }

    override fun copyFile(requestor: Any?, virtualFile: VirtualFile, newParent: VirtualFile, copyName: String): VirtualFile {
        throw IncorrectOperationException()
    }

    override fun isReadOnly(): Boolean = true

    companion object {
        @NonNls
        const val PROTOCOL: String = "methods"

        fun getInstance(): MethodsVirtualFileSystem =
            VirtualFileManager.getInstance().getFileSystem(PROTOCOL) as MethodsVirtualFileSystem
    }
}

